subprojects/docs/src/docs/userguide/dependency_management_attribute_based_matching.adoc

```diff
@@ -170,3 +170,134 @@ Because multiple values for an attribute can be _compatible_ with the requested
+include::sample[dir="userguide/dependencyManagement/artifactTransforms/unzip/kotlin",files="build.gradle.kts[tags=artifact-transform-registration]"]
+====
+
+After that registration, every time a module with artifact type `java-classes-directory` is requested, but the module only has a variant of type `jar`, the `jar` is unpacked and the directory with the classes is provided to the resolved configuration.
+
+Note that artifact transforms are executed in parallel by default.
+
+=== Registering Artifact Transforms
+
+For registering an artifact transform, the method link:{groovyDslPath}/org.gradle.api.artifacts.dsl.DependencyHandler.html#org.gradle.api.artifacts.dsl.DependencyHandler:registerTransform(java.lang.Class,%20org.gradle.api.Action)[dependencies.registerTransform] is used.
+As part of the registration, the `from` and `to` attributes need to be supplied.
+When required, parameters for the transform can be configured.
+
+The artifact transform needs to be registered in the project which consumes the transformed artifacts.
+
+.Artifact transform registration with parameters
+====
+include::sample[dir="userguide/dependencyManagement/artifactTransforms/relocate/groovy",files="build.gradle[tags=artifact-transform-registration]"]
+include::sample[dir="userguide/dependencyManagement/artifactTransforms/relocate/kotlin",files="build.gradle.kts[tags=artifact-transform-registration]"]
+====
+
+=== Artifact Transform Selection
+
+Depending on the producer attributes and the consumer attributes Gradle decides which transforms need to be executed on a given artifact.
+It starts with the requested attributes and then considers all transforms which modify some of the requested attributes as possible paths leading there.
+Going backwards, Gradle tries to obtain a path to some producer variant using transforms.
+Of all the found transform chains, Gradle tries to select the best one:
+
+- If there is only one transform chain, it is selected.
+- If there are two transform chains, and one is a suffix of the other one, it is selected.
+- If there is a shortest transform chain, then it is selected.
+- In all other cases, the selection fails and an error is reported.
+
+Note that artifact transforms are not considered when there is already a producer variant matching the requested attributes.
+
+[NOTE]
+====
+The `artifactType` attribute is special, since it is only present on artifacts and not on modules.
+As a consequence, any transform which is only mutating `artifactType` will never be selected when transforming between configurations with different consumer and producer attributes.
+It will only be considered when using an link:{javadocPath}/org/gradle/api/artifacts/ArtifactView.html[ArtifactView].
+====
+
+=== Implementing Artifact Transforms
+
+Similar to tasks, an artifact transform consists of an action and some parameters.
```

Comment:
We want to move away from `properties` and move to using `parameters`. For artifact transforms we are already there, since the method to configure the parameters is called `parameters`. For tasks, there is no separation between the task type and the parameter type, so the parameters live on the task type as properties.